ssl/ssl_lib.c

```diff
@@ -5847,7 +5847,7 @@ int SSL_stateless(SSL *s)
     ret = SSL_accept(s);
     s->s3.flags &= ~TLS1_FLAGS_STATELESS;
 
-    if (ret > 0 && s->ext.cookieok)
+    if (ret > 0 && (s->ext.cookieok || !SSL_IS_TLS13(s)))
```

Comment:
but the TLS connection is fine, the only problem is the return value of SSL_stateless is -1 and thus our
s_server command closes the connection after SSL_stateless returns -1.

Reply:
Furthermore there is no way how SSL_stateless can work with the -nbio option.  I added a check for that.
And I found out that the s_server's -nbio_test option is not working since the big bang, fixed that too.

Reply:
>This does not seem right to me. From the docs for SSL_stateless():


I wouldn't  care too much about the docs here.
What SSL_stateless does is forcing TLS1.3 clients to do a HRR,
That is done in a stateless way, and it is NOT working with non-blocking sockets.
Then it internally calls SSL_accept, but that does NOT work with non-blocking
sockets either and this is not done in a stateless way either.
So the SSL_accept will just wait for the Client to send the
Change Cipher spec and so on.

Reply:
As Matt notes, we don't expect any sort of ChangeCipherState for a pure TLS 1.3 connection.  We might, however, see one if the "middlebox compatibility mode" (https://datatracker.ietf.org/doc/html/rfc8446#appendix-D.4) is in use.
Given that `SSL_stateless()` is currently of extremely limited utility, I would be willing to update the documentation to note that it is incompatible with clients using middlebox compatibility mode.

Reply:
It is the default setting **for OpenSSL clients**, which hardly constitute the entire TLS ecosystem